### PR TITLE
NUMBERS-188: refactored Complex instance methods as static functions

### PR DESCRIPTION
**Pull request by sumanth-rajkumar**
This PR refactors Complex class using functional interfaces and static methods. 
This allows reuse of the refactored static functions for Operations on List/Matrices of Complex Numbers (NUMBERS-186)

The changes retain binary backward compatibility.

Summary of changes

1) Introduced following interfaces  for Complex Number Operations
 ComplexDouble - interface representing a complex number with double precision in cartesian form
 ComplexUnaryOperator - unary operations on ComplexDouble 
 ComplexBinaryOperator - binary operations on ComplexDouble 
 ComplexScalarFunction -  operations on a ComplexDouble and double type returning a ComplexDouble type
 ComplexConstructor - interface to create the ComplexDouble result object from real and imaginary parts

2) Refactored instance methods of Complex class as static functions in ComplexFunctions and ComplexBiFunctions
The static functions use the functional interface signatures described above

3) Added default methods on functional interfaces to compose Complex operations.

### CODE REVIEW

**Comment by aherbert** [comment]
Currently `complex-gsoc-2022` is up-to-date with `master` so you can rebase on that branch. I have added changes to detect missing javadoc on private methods. Please rebase to collect the changes. Thanks.

**Comment by sumanth-rajkumar** [comment]
> Currently `complex-gsoc-2022` is up-to-date with `master` so you can rebase on that branch. I have added changes to detect missing javadoc on private methods. Please rebase to collect the changes. Thanks.

**Comment by sumanth-rajkumar** [comment]
Hi Alex,

I had taken out comments of the private methods in ComplexFunctions and ComplexBiFunctions because I had the error of the class being over 2000 lines. Is there a way around this?

**Comment by aherbert** [comment]
You should add a suppression to:
```
src/main/resources/checkstyle/checkstyle-suppressions.xml
```
There is already this suppression for Complex:
```
<suppress checks="FileLengthCheck" files=".*[/\\]Complex(Test)?\.java" />
```
You may be able to drop the suppression for Complex if it is now under 2000 lines and change it to your other files.

If you rebase on the branch again (because I have now updated the notifications for GH actions to send e-mails upon failure) and I can then approve your PR to allow it to run the GH workflows.

**Comment by sumanth-rajkumar** [comment]
Based on feedback on the mailing list, I have made the following changes

1) Rebased and added missing Javadoc comments

2) Removed abstractions for simple functions in the Complex class

3) Fixed line space formatting

4) Renamed the interface DComplex as ComplexDouble 

Could you please approve the PR to run the GH workflows?

File: commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexScalarFunction.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    /**
+     * Returns a composed scalar function that first applies this function to its input, and then applies the after function to the result.
+     * If evaluation of either function throws an exception, it is relayed to the caller of the composed function.
+     * @param after the function to apply after this function is applied
+     * @return a composed function that first applies this function and then applies the after function
+     */
+    default ComplexScalarFunction thenApply(ComplexUnaryOperator after) {
+        Objects.requireNonNull(after);
+        return (ComplexDouble c, double f, ComplexConstructor<ComplexDouble> out) -> after.apply(apply(c, f, out), out);
+
```
Comment (aherbert) [inline commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexScalarFunction.java:47]: Please check the code for these redundant additional whitespace lines.

File: commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexUnaryOperator.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        return apply(c, ComplexConstructor.D_COMPLEX_RESULT);
+    }
+
+    /**
+     * Returns a composed unary operator that first applies this function to its input, and then applies the after UnaryOperator function to the result.
+     * If evaluation of either function throws an exception, it is relayed to the caller of the composed function.
+     * @param after the function to apply after this function is applied
+     * @return a composed function that first applies this function and then applies the after function
+     */
+    default ComplexUnaryOperator thenApply(ComplexUnaryOperator after) {
```
Comment (aherbert) [inline commons-numbers-complex/src/main/java/org/apache/commons/numbers/complex/ComplexUnaryOperator.java:65]: Following `java.util.function.Function` this should have a `compose` and a `andThen`.

There is no precedent in the JDK functions for composites that expand the argument list (making a unary operator into a binary operator), i.e. the methods `thenApplyBinaryOperator` and `thenApplyScalarFunction`. Can these be removed?

I note you have used them in the pow functions. But these are not required in the public API.

**Review by aherbert — CHANGES_REQUESTED** [review]
Thanks for the code changes. I've provided some initial items to address. Feel free to raise points on dev ML for further clarification/discussion.

File: commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/CStandardTest.java [hunk 3]
```diff
@@ -865,12 +884,29 @@ void testImplicitTrig() {
             final double re = next(rng);
             final double im = next(rng);
             final Complex z = complex(re, im);
-            final Complex iz = z.multiplyImaginary(1);
-            assertComplex(z.asin(), iz.asinh().multiplyImaginary(-1));
-            assertComplex(z.atan(), iz.atanh().multiplyImaginary(-1));
-            assertComplex(z.cos(), iz.cosh());
-            assertComplex(z.sin(), iz.sinh().multiplyImaginary(-1));
-            assertComplex(z.tan(), iz.tanh().multiplyImaginary(-1));
+            final Complex iz1 = z.multiplyImaginary(1);
```
Comment (aherbert) [inline commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/CStandardTest.java:887]: The code repetition here is prone to error. The operations should be tested using TestUtils.assert with two operations created using composition:
```Java
ComplexUnaryOperator<ComplexNumber> mi1 = (x, y, out) -> ComplexFunctions.multiplyImaginary(x, y, 1, out);
ComplexUnaryOperator<ComplexNumber> mim1 = (x, y, out) -> ComplexFunctions.multiplyImaginary(x, y, -1, out);

UnaryOperator<Complex> asin1 = c -> c.multiplyImaginary(1).asinh().multiplyImaginary(-1);
ComplexUnaryOperator<ComplexNumber> asin2 = mi1.andThen(ComplexFunctions::asinh).andThen(mim1);

// Test the two functions produce the same result
// Test it is the expected result
```

File: commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexComposeTest.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    private static final ComplexUnaryOperator<ComplexDouble> neg = ComplexFunctions::negate;
+    private static final ComplexUnaryOperator<ComplexDouble> multiplyImag = ComplexFunctions::multiplyImaginary;
+    private static final ComplexUnaryOperator<ComplexDouble> conj = ComplexFunctions::conj;
+    private static final ComplexUnaryOperator<ComplexDouble> multiplyImagConj = multiplyImag.andThen(conj);
+    private static final ComplexUnaryOperator<ComplexDouble> conjMultiplyImag = conj.andThen(multiplyImag);
+    private static final ComplexUnaryOperator<ComplexDouble> identity1 = multiplyImagConj.andThen(multiplyImagConj);
+    private static final ComplexUnaryOperator<ComplexDouble> identity2 = conjMultiplyImag.andThen(conjMultiplyImag);
+
+    @Test
+    void testConjugateIdentities() {
```
Comment (aherbert) [inline commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexComposeTest.java:34]: This type of test is better as a `@ParameterizedTest` using a `@CsvSource` to provide the (r, i) pairs. Since the same numbers are used for two tests you could use a `@MethodSource` for the two test methods that require a single number:
```Java

    static Stream<Arguments> singleComplex() {
        return Stream.of(
            Arguments.of(3, 4),
            Arguments.of(3, -4)
        );
    }

    @ParameterizedTest
    @MethodSource(value = {"singleComplex"})
    void testConjugateIdentity(double r, double i) {
        Complex c = Complex.ofCartesian(r, i);

```

File: commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexComposeTest.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.commons.numbers.complex;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+class ComplexComposeTest {
```
Comment (aherbert) [inline commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexComposeTest.java:23]: Javadoc would be helpful, e.g. What is the overall purpose of this class?

File: commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexFunctionsTest.java [hunk 6]
```diff
@@ -0,0 +1,1175 @@
...
+ *
+ * <ul>
+ * <li>{@link CStandardTest} for a test of the ISO C99 standards including special case handling.
+ * <li>{@link CReferenceTest} for a test of the output using standard finite value against an
+ *     ISO C99 compliant reference implementation.
+ * <li>{@link ComplexEdgeCaseTest} for a test of extreme edge case finite values for real and/or
+ *     imaginary parts that can create intermediate overflow or underflow.
+ * </ul>
+ */
+class ComplexFunctionsTest {
```
Comment (aherbert) [inline commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/ComplexFunctionsTest.java:42]: Is this entire class redundant?

If you update ComplexTest to test all current functions using the Complex and ComplexFunctions methods with TestUtils.assert methods then you have already covered all the functionality.

It means that any updated to ComplexTest do not have to be duplicated here.

**Review by aherbert — CHANGES_REQUESTED** [review]
Hi @sumanth-rajkumar 

I think this needs to be rebased on the latest branch changes. Your diff has changes to files that are already in the branch.

This PR requires some more work. All the tests on the Complex methods should use the TestUtils assertions. These have the complex method and the corresponding ComplexFunctions methods. The test should assert the two operations are binary equal. Then they should assert the result is binary equal to the expected (assertSame) or within tolerance (assertEquals). This binary equality is very important for complex numbers that have zeros as the sign is part of the result.

I've provided comments throughout the code but not marked everything as there are themes that should be generally applied.

Please render the javadoc for complex functions and the new functional interfaces. Then read it. This forces you to read the description as the code is not available. The javadoc should describe the API functionality and not require access to the source code to check what the methods/interfaces/classes are for.

File: commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/TestUtils.java [hunk 7]
```diff
@@ -387,4 +424,314 @@ private static String preprocessTestData(String line, TestDataFlagOption option,
...
+     * @param msg the failure message
+     * @param expected the expected
+     * @param actual the actual
+     * @param delta delta
+     */
+    static void assertEquals(Supplier<String> msg, double expected, double actual, double delta) {
+        Assertions.assertEquals(expected, actual, delta, msg);
+    }
+
+    static class ComplexDoubleConstructor implements ComplexConstructor<ComplexDouble>, ComplexDouble {
```
[the remark this answers is not in the document]
Comment (sumanth-rajkumar) [reply commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/TestUtils.java:692]: Is there a way I can call the ComplexNumber class in the assertComplex operations in the other test classes? Before I said TestUtils.ComplexDoubleConstructor.of() in the apply operation, is there a way to do it like this with your suggested ComplexNumber class

[the remark this answers is not in the document]
Comment (aherbert) [reply commons-numbers-complex/src/test/java/org/apache/commons/numbers/complex/TestUtils.java:692]: The class is package private. You should just have to import ` ...TestUtils.ComplexNumber` to use it elsewhere. Or move the ComplexNumber to a new class in the test package to make it clear it is for reuse.

**Comment by aherbert** [comment]
This work has been superseded by other PRs merged into the branch to apply the changes in incremental stages.
